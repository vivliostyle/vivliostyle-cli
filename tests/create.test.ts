import './mocks/bluwy__giget-core.js';
import './mocks/fs.js';
import './mocks/tmp.js';

import { vol } from 'memfs';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import {
  TEMPLATE_ANSWER_NOT_USE,
  THEME_ANSWER_MANUAL,
  THEME_ANSWER_NOT_USE,
} from '../src/core/create.js';
import type { PackageJson, PackageSearchResult } from '../src/npm';
import { runCommand } from './command-util';

const mockedClackModule = vi.hoisted(() => {
  const mockedAnswers = vi.fn().mockReturnValue({});
  const prompt = vi
    .fn()
    .mockImplementation(async ({ name }: { name: string }) => {
      const answers = mockedAnswers();
      if (!(name in answers)) {
        throw new Error(`Unexpected question: ${name}`);
      }
      return answers[name];
    });

  return {
    answers: mockedAnswers,
    text: prompt,
    select: prompt,
    multiSelect: prompt,
    autocomplete: prompt,
    autocompleteMultiselect: prompt,
    isCancel: vi.fn().mockReturnValue(false),
    log: {
      warn: vi.fn().mockImplementation((message: string) => {
        throw new Error(`Unexpected call to log.warn: ${message}`);
      }),
    },
  };
});

vi.mock('@clack/prompts', () => mockedClackModule);

const mockedNpmModule = vi.hoisted(async () => {
  const fs = await import('node:fs');
  const path = await import('node:path');
  const { fileURLToPath } = await import('node:url');

  const testPackage = JSON.parse(
    fs.readFileSync(
      path.resolve(
        fileURLToPath(import.meta.url),
        '../fixtures/themes/theme-with-template/package.json',
      ),
      'utf-8',
    ),
  ) as PackageJson;
  return {
    createFetch: () =>
      (() => {}) as unknown as typeof import('node-fetch-native').fetch,
    listVivliostyleThemes: () =>
      Promise.resolve({
        objects: [
          {
            downloads: { monthly: 100, weekly: 25 },
            package: testPackage,
            searchScore: 1,
            updated: 'time',
          },
        ],
        total: 1,
        time: 'time',
      } satisfies PackageSearchResult),
    fetchPackageMetadata: ({ packageName }) =>
      Promise.resolve(
        packageName === '@vivliostyle/custom-theme'
          ? {
              name: '@vivliostyle/custom-theme',
              version: '9.9.9',
            }
          : testPackage,
      ),
  } satisfies typeof import('../src/npm');
});

vi.mock('../src/npm', () => mockedNpmModule);

beforeEach(() => {
  vol.reset();
  mockedClackModule.answers.mockReturnValue({});
});

describe('create command', () => {
  it('create empty project', async () => {
    mockedClackModule.answers.mockReturnValue({
      projectPath: 'project-name',
      title: 'Booook titleeee',
      author: 'Authoooor',
      language: 'en',
      theme: THEME_ANSWER_NOT_USE,
    });

    await runCommand(['create'], { cwd: '/work' });
    const files = vol.toJSON();
    delete files['/work/project-name/manuscript.md'];
    expect(files).toMatchSnapshot();
  });

  it('create project with a default template', async () => {
    mockedClackModule.answers.mockReturnValue({
      projectPath: 'project-name',
      title: 'Booook of the titleeee',
      author: 'Authoooor',
      language: 'en',
      theme: 'theme-with-template',
      usingTemplate: 'default',
    });

    await runCommand(['create'], { cwd: '/work' });
    const files = vol.toJSON();
    expect(files).toMatchSnapshot();
  });

  it('create project with a custom-prompt template', async () => {
    mockedClackModule.answers.mockReturnValue({
      projectPath: 'project-name',
      title: 'Booook titleeee',
      author: 'Authoooor',
      language: 'en',
      theme: 'theme-with-template',
      usingTemplate: 'custom-prompt',
      stringPromptA: 'custom string',
      selectPromptA: 'option1',
    });

    await runCommand(['create'], { cwd: '/work' });
    const files = vol.toJSON();
    expect(files).toMatchSnapshot();
  });

  it('create project with a custom theme', async () => {
    mockedClackModule.answers.mockReturnValue({
      projectPath: 'project-name',
      title: 'Booook titleeee',
      author: 'Authoooor',
      language: 'en',
      theme: THEME_ANSWER_MANUAL,
      themeManualInput: '@vivliostyle/custom-theme',
      usingTemplate: TEMPLATE_ANSWER_NOT_USE,
    });

    await runCommand(['create'], { cwd: '/work' });
    const files = vol.toJSON();
    expect(files['/work/project-name/vivliostyle.config.js']).toMatch(
      'theme: "@vivliostyle/custom-theme@^9.9.9"',
    );
  });

  it('avoid overwrite', async () => {
    vol.fromJSON({
      '/work/out/touch': '',
    });
    mockedClackModule.answers.mockReturnValue({
      projectPath: 'out',
    });

    await expect(runCommand(['create'], { cwd: '/work' })).rejects.toThrow(
      'Destination /work/out is not empty.',
    );
  });
});

describe('init command', () => {
  it('test the init command', async () => {
    await runCommand(['init'], { cwd: '/work' });
    const files = vol.toJSON();
    expect(files['/work/vivliostyle.config.js']).toMatchSnapshot();
  });

  it('test the init command with long options', async () => {
    await runCommand(
      [
        'init',
        '--title',
        'Sample Document',
        '--author',
        'Author Name <author@example.com>',
        '--language',
        'ja',
        '--size',
        'A5',
        '--theme',
        'style.css',
      ],
      { cwd: '/work' },
    );
    const files = vol.toJSON();
    expect(files['/work/vivliostyle.config.js']).toMatchSnapshot();
  });

  it('test the init command with short options', async () => {
    await runCommand(
      [
        'init',
        '--title',
        'Sample Document2',
        '--author',
        'Author Name2 <author@example.com>',
        '-l',
        'pt-BR',
        '-s',
        'A3',
        '-T',
        'theme.css',
      ],
      { cwd: '/work' },
    );
    const files = vol.toJSON();
    expect(files['/work/vivliostyle.config.js']).toMatchSnapshot();
  });
});
