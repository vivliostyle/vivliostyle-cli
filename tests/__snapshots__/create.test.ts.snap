// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create command > create empty project 1`] = `
Object {
  "/work/project-name/.gitignore": "node_modules/
.vivliostyle/
",
  "/work/project-name/README.md": "# Booook Titleeee

## References

- VFM <https://vivliostyle.github.io/vfm/#/vfm>
- Vivliostyle CLI <https://github.com/vivliostyle/vivliostyle-cli#readme>
- Vivliostyle Themes <https://github.com/vivliostyle/themes#readme>
- Awesome Vivliostyle <https://github.com/vivliostyle/awesome-vivliostyle#readme>
- Vivliostyle (GitHub) <https://github.com/vivliostyle>
- Vivliostyle <https://vivliostyle.org>
",
  "/work/project-name/package.json": "{
  \\"name\\": \\"booook-titleeee\\",
  \\"description\\": \\"Booook Titleeee\\",
  \\"author\\": \\"Authoooor\\",
  \\"version\\": \\"0.0.0\\",
  \\"type\\": \\"module\\",
  \\"private\\": true,
  \\"scripts\\": {
    \\"build\\": \\"vivliostyle build\\",
    \\"preview\\": \\"vivliostyle preview\\"
  },
  \\"dependencies\\": {
    \\"@vivliostyle/cli\\": \\"0.0.1\\"
  }
}
",
  "/work/project-name/vivliostyle.config.js": "/** @type {import('@vivliostyle/cli').VivliostyleConfigSchema} */
module.exports = {
  title: \\"Booook Titleeee\\",
  author: \\"Authoooor\\",
  image: \\"ghcr.io/vivliostyle/cli:0.0.1\\",
  entry: [\\"manuscript.md\\"],
};
",
}
`;

exports[`create command > create project with a custom-prompt template 1`] = `
Object {
  "/work/project-name/manuscript.md": "# Custom Prompt Custom String

You chosen option1
",
  "/work/project-name/vivliostyle.config.js": "/** @type {import('@vivliostyle/cli').VivliostyleConfigSchema} */
export default {
  title: 'Booook titleeee',
  entry: 'manuscript.md',
};
",
}
`;

exports[`create command > create project with a default template 1`] = `
Object {
  "/work/project-name/default.md": "# Booook of the titleeee

BOOOOK OF THE TITLEEEE
booook of the titleeee
Booook Of The Titleeee
booookOfTheTitleeee
booook_of_the_titleeee
booook-of-the-titleeee
Booook of the Titleeee

projectPath: project-name
title: Booook of the titleeee
author: Authoooor
themePackage: theme-with-template@1.0.0
cliVersion: 0.0.1
coreVersion: 0.0.1
",
  "/work/project-name/vivliostyle.config.js": "/** @type {import('@vivliostyle/cli').VivliostyleConfigSchema} */
export default {
  title: 'project-name',
  entry: 'default.md',
};
",
}
`;

exports[`init command > test the init command 1`] = `
"/** @type {import('@vivliostyle/cli').VivliostyleConfigSchema} */
module.exports = {
  title: \\"My Book Title\\",
  author: \\"John Doe\\",
  image: \\"ghcr.io/vivliostyle/cli:0.0.1\\",
  entry: [\\"manuscript.md\\"],
};
"
`;

exports[`init command > test the init command with long options 1`] = `
"/** @type {import('@vivliostyle/cli').VivliostyleConfigSchema} */
module.exports = {
  title: \\"Sample Document\\",
  author: \\"Author Name <author@example.com>\\",
  language: \\"en\\",
  size: \\"A5\\",
  theme: \\"style.css\\",
  image: \\"ghcr.io/vivliostyle/cli:0.0.1\\",
  entry: [\\"manuscript.md\\"],
};
"
`;

exports[`init command > test the init command with short options 1`] = `
"/** @type {import('@vivliostyle/cli').VivliostyleConfigSchema} */
module.exports = {
  title: \\"Sample Document2\\",
  author: \\"Author Name2 <author@example.com>\\",
  language: \\"jp\\",
  size: \\"A3\\",
  theme: \\"theme.css\\",
  image: \\"ghcr.io/vivliostyle/cli:0.0.1\\",
  entry: [\\"manuscript.md\\"],
};
"
`;
